File: java/Matrix.java
/* Calculadora de matrices
 * Para definir la matriz principal Matrix matrizPrincipal = new Matrix(int[][])
 * Para sumar dos matrices matrizPrincipal.suma(<Matriz sumando>, operador) (Matriz sumando = int[][]) (operador<Char> -> '+' o '-')
 * Para multiplicar dos matrices matrixPrincipal.multiplicar(matrizB) (matrizB = int[][])
 * Para calcular el cuadrado de la matriz. matrizPrincipal.square()
 * Para multiplicar una matriz por un número -> matrizPrincipal.scalar(<Integer>)
 * En el ejemplo aportado solo se definen dos matrices multiplicables
 */

public class Matrix {
    protected int[][] matrix;

    // Código principal que se ejecutará
    public static void main(String[] args) {
        int[][] matriz = {
                { 1, 2, 1 },
                { 2, 0, 1 },
        };

        int[][] multiplicador = {
                { 1, 2 },
                { 2, 1 },
                { 0, 1 }
        };

        Matrix m = new Matrix(matriz); // Crear la matriz sobre la que se trabajará
        m.scalar(2);
    }

    // Constructor de la class Matrix
    public Matrix(int[][] mat) {
        this.matrix = mat; // Guardar en la variable matrix la matriz introducida por el usuario
    }

    // Escribir la matriz en la consola, ya que por defecto se genera un código en
    // lugar de la representación deseada
    public void WriteMatrix() {
        for (int i = 0; i < matrix.length; i++) {
            for (int j = 0; j < matrix[0].length; j++) {
                System.out.print(matrix[i][j]); // Escribir cada linea de la matriz
                System.out.print("  "); // Separador de números para identificar bien cada elemento
            }
            System.out.println(); // Al acabar la fila de la matriz hacer un salto de línea
        }
    }

    public void scalar(int scalar) {
        for(int i = 0; i<matrix.length; i++) {
            for(int j = 0; j<matrix[0].length; j++) {
                matrix[i][j] *= scalar;
            }
        }

        WriteMatrix();
    }
    // Sumar dos matrices
    public void suma(int[][] matrizB, char operador) {
        int rows = matrix.length; // Almacenar las filas de la matriz "principal"
        int cols = matrix[0].length; // Almacenar las columnas de la matriz "principal"

        int oRows = matrizB.length; // Almacenar las filas de la matriz que se va a sumar / restar
        int oCols = matrizB[0].length; // Almacenar las columnas de la matriz que se va a sumar / restar

        if (rows != oRows || cols != oCols) {
            return;
        } // Comprobar que las matrices sean de las mismas dimensiones, si no lo son no se
          // pueden sumar

        if (operador == '+') {
            for (int i = 0; i < rows; i++) {
                for (int j = 0; j < cols; j++) {
                    matrix[i][j] += matrizB[i][j]; // Sumar a cada elemento de la matriz "principal" el elemento que le
                                                   // corresponde en la matriz sumando, y guardar su valor
                }
            }
        } else if (operador == '-') {
            for (int i = 0; i < rows; i++) {
                for (int j = 0; j < cols; j++) {
                    matrix[i][j] -= matrizB[i][j];
                }
            }
        } else {
            System.out.println("Las operaciones válidas son '+' y '-'");
        }

        WriteMatrix(); // Mostrar la matriz en la consola
    }

    public void multiplicar(int[][] multiplicador) {
        int rows = matrix.length; // Numero de filas de la matriz "principal"
        int cols = matrix[0].length; // Número de columnas de la matriz "principal"

        int oRows = multiplicador.length; // Número de filas de la matriz por la que se va a multiplicar
        int oCols = multiplicador[0].length; // Número de columnas de la matriz por la que se va a multiplicar

        if (cols != oRows) {
            return;
        } // Comprobar si el número de columnas de la matriz tiene el mismo número de
          // filas que la segunda matriz, de modo que sean multiplicables

        int[][] result = new int[rows][oCols];

        // Multiplicar la matriz
        for (int i = 0; i < rows; i++) {
            result[i] = new int[oCols];
            for (int j = 0; j < oCols; j++) {
                result[i][j] = 0;
                for (int k = 0; k < cols; k++) {
                    result[i][j] += matrix[i][k] * multiplicador[k][j];
                }
            }
        }

        matrix = result;

        WriteMatrix(); // Mostrar la matriz en la consola
    }

    public void square() {
        multiplicar(matrix);
    }
}

File: js/matrix.js
/*
    COMO USAR:

    1º Crear una matriz (const miMatriz = new Matrix(<Array>))

    UTILIDADES:

    Añadir una fila a la matriz -> miMatriz.addRow(<Array>)
    Añadir una columna a la matriz -> miMatriz.addColumn(<Array>)
    Multiplicar la matriz por una constante -> miMatriz.scalarMult(<Number>)
    Sumar 2 matrices -> miMatriz.sum(<Matriz>, <String>) donde <String> debe ser '+' o '-'
    Multiplicar 2 matrices -> miMatriz.multiply(<Matriz>)
    Calcular el cuadrado de la matriz -> miMatriz.square()

    (En todos los casos se compueba que la operación sea realizable, no es necesario asegurarlo "a mano")

    LAS MATRICES PUESTAS DE EJEMPLO SON MULTIPLICABLES, PERO NO SUMABLES
*/

class Matrix {
    constructor(mat) {
        this.matrix = mat
    }

    addRow(arr) {
        if (this.matrix[0].length != arr.length) {
            console.log("La fila debe contener el mismo número de números que las otras filas")
            return
        }
        this.matrix.push(arr) //Para añadir una fila a nuestra matriz sólo hace falta añadir un array al final de ella
    }

    //Para añadir una columna hay que añadir al final de cada línea el valor de la columna en la posición de la fila
    addColumn(arr) {
        let rows = this.matrix.length
        let cols = arr.length;

        if (rows != cols) { console.log("Debes poner una columna que tenga la misma cantidad de números que las demás columas"); return; }

        for (let i = 0; i < rows; i++) {
            this.matrix[i].push(arr[i])
        }
    }

    //Para multiplicar una matriz por un número hay que multiplicar todos los elementos de la misma por ese número
    scalarMult(scalar) {
        let rows = this.matrix.length;
        let cols = this.matrix[0].length;

        for (let i = 0; i < rows; i++) {
            for (let j = 0; j < cols; j++) {
                this.matrix[i][j] *= scalar
            }
        }
    }

    //Para poder sumar dos matrices estas deben tener las mismas dimensiones. Hay que sumar cada numero con aquel que esté en la misma posición de la otra matriz
    sum(mat, op) {
        //Filas y columnas de nuestra matriz
        let rows = this.matrix.length;
        let cols = this.matrix[0].length;

        //Filas y columnas de la otra matriz
        let oRows = mat.length;
        let oCols = mat[0].length;

        //Se comprueba si la dimensión de las dos matrices es la misma
        if (rows != oRows || cols != oCols) {
            console.log('LAS MATRICES NO SON SUMABLES')
            return
        }

        //Si es suma, se suman
        if (op == '+') {
            for (let i = 0; i < rows; i++) {
                for (let j = 0; j < cols; j++) {
                    this.matrix[i][j] += mat[i][j]
                }
            }
        } else if (op == '-') {
            //Si es una resta se restan
            for (let i = 0; i < rows; i++) {
                for (let j = 0; j < cols; j++) {
                    this.matrix[i][j] -= mat[i][j]
                }
            }
        } else {
            //Si no es ni una suma ni una resta, se descarta la operación
            console.log('LA OPERACIÓN NO ESTÁ PERMITIDA')
        }
    }

    multiply(mat) {
        let rows = this.matrix.length; //Filas de esta matriz
        let cols = this.matrix[0].length; //Columnas de esta matriz
        let oRows = mat.length //Filas de la otra matriz
        let oCols = mat[0].length //Columnas de la otra matriz

        //Para que una matriz pueda ser multiplicada por otra el número de columnas de la primera debe ser igual al número de filas de la segunda
        if (cols != oRows) {
            console.log('LAS MATRICES NO PUEDEN SER MULTIPLICADAS')
            return
        }

        const result = new Array(rows); //Array donde se irán guardando los valores según se vayan calculando

        //Multiplicar cada numero (fila * columna) y sumarlos entre ellos -> Ponerlos en la posición del Array temporal en la posición en la que corresponda
        for (let i = 0; i < rows; i++) {
            result[i] = new Array(oCols);
            for (let j = 0; j < oCols; j++) {
                result[i][j] = 0;
                for (let k = 0; k < cols; k++) {
                    result[i][j] += this.matrix[i][k] * mat[k][j];
                }
            }
        }

        //Actualizar la matriz original con el resultado de la matriz nueva
        this.matrix = result
    }
    square() {
        this.multiply(this.matrix) //Multiplicar la matriz por si misma
    }
}

//Esta matriz es solo una de demostración, para usar todas las funcionalidades se deben crear más matrices, que además deben ser las apropiadas para cada tipo de operación
let matriz = [
    [1, 2, 1],
    [2, 0, 1],
    [1, 2, 0]
]

//Crear una instacia de la matriz
const m = new Matrix(matriz)
m.addRow([1, 2])

console.log(m)
